src/lib/pbo/quadratization.jl

```diff
@@ -1,205 +1,178 @@
-        # Without any extra knowledge, it is better to
-        # quadratize term-by-term
-        return Quadratization{TBT}(k)
-    end
-end
-
-@doc raw"""
-    quadratize(aux::Function, f::PBF{S, T}, ::Quadratization) where {S, T}
-
-Quadratizes a given PBF, i.e. creates a function ``g \in \mathscr{F}^{2}`` from ``f \in \mathscr{F}^{k}, k \ge 3``.
+Quadratizes a given PBF in-place, i.e. applies a mapping ``Q : \mathscr{F}^{k} \to \mathscr{F}^{2}``.
 
 ```julia
 aux(::Nothing)::S
 aux(::Integer)::Vector{S}
 ```
-
-## Submodularity
-
-A function ``f : 2^{S} \to \mathbb{R}`` is said to be submodular if
-```math
-f(X \cup Y) + f(X \cap Y) \le f(X) + f(Y) \forall X, Y \subset S
-```
-""" function quadratize end
+""" function quadratize! end
 
 @doc raw"""
-    Quadratization{TBT}(::Int)
+    Quadratization{NTR_KZFD}(stable::Bool = false)
 
-Term-by-term quadratization. Employs other inner methods.
-"""
+NTR-KZFD (Kolmogorov & Zabih, 2004; Freedman & Drineas, 2005)
 
-@quadratization(TBT, 0, 0)
+!!! info
+    Introduces one new variable and no non-submodular terms.
+""" struct NTR_KZFD <: QuadratizationMethod end
+
+function quadratize!(
+    aux::Function,
+    f::PBF{S,T},
+    ω::Set{S},
+    c::T,
+    ::Quadratization{NTR_KZFD},
+) where {S,T}
+    # -* Degree *- #
+    k = length(ω)
 
-@doc raw"""
-    Quadratization{NTR_KZFD}(::Int)
+    # -* Fast-track *- #
+    k < 3 && return nothing
 
-NTR-KZFD (Kolmogorov & Zabih, 2004; Freedman & Drineas, 2005)
-"""
+    # -* Variables *- #
+    s = aux()::S
 
-@quadratization(NTR_KZFD, 1, 0)
+    # -* Stabilize *- #
+    # NOTE: This method is stable by construction
 
-function quadratize(aux::Function, ω::Set{S}, c::T, ::Quadratization{NTR_KZFD}) where {S, T}
-    # -* Degree *-
-    k = length(ω)
-    s = aux()::S
+    # -*- Quadratization -*- #
+    delete!(f, ω)
 
-    g = PBF{S,T}(s => -c * (k - 1))
+    f[s] -= c * (k - 1)
 
     for i ∈ ω
-        η = Set{S}([s, i])
-
-        g[η] += c
+        f[i×s] += c
     end
 
-    return g
+    return nothing
 end
 
 @doc raw"""
-    Quadratization{PTR_BG}(::Int)
+    Quadratization{PTR_BG}(stable::Bool = false)
 
 PTR-BG (Boros & Gruber, 2014)
-"""
-
-@quadratization(
-    PTR_BG,
-    k -> k - 2,
-    k -> k - 1,
-)
 
-function quadratize(aux::Function, ω::Set{S}, c::T, ::Quadratization{PTR_BG}) where {S, T}
-    # -* Degree *-
+!!! info
+    Introduces ``k - 2`` new variables and ``k - 1`` non-submodular terms.
+""" struct PTR_BG <: QuadratizationMethod end
+
+function quadratize!(
+    aux::Function,
+    f::PBF{S,T},
+    ω::Set{S},
+    c::T,
+    quad::Quadratization{PTR_BG},
+) where {S,T}
+    # -* Degree *- #
     k = length(ω)
 
-    # -* Variables *-
+    # -* Fast-track *- #
+    k < 3 && return nothing
+
+    # -* Variables *- #
     s = aux(k - 2)::Vector{S}
-    b = sort(collect(ω); lt = varcmp)::Vector{S}
+    b = collect(ω)::Vector{S}
+
+    # -* Stabilize *- #
+    quad.stable && sort!(b; lt = varlt)
+
+    # -*- Quadratization -*- #
```

Comment:
we should unifor comments.
all should be simple like `# Stabilize` and `# Quadratization`

Reply:
I went through the most of it. Maybe there are still a few ones to be fixed but not many.